tests/h/settings_test.py

```diff
@@ -3,98 +3,104 @@
 
+    def test_sets_value_when_deprecated_setting_is_used(self):
+        settings_manager = settings.SettingsManager({}, {'FOO': 'bar'})
+        settings_manager.set('foo', 'FOO', deprecated_msg='what to do instead')
 
-def asutf8(str):
-    return str.encode()
+        result = settings_manager.settings['foo']
 
+        assert result == 'bar'
 
-@pytest.mark.parametrize('setting,varname,type,environ,expected', (
-    # Should return None when the env var in question isn't set
-    ('foo', 'FOO', None, {}, None),
+    def test_warns_when_deprecated_setting_is_used(self, caplog):
+        with caplog.at_level(logging.WARN):
+            settings_manager = settings.SettingsManager({}, {'FOO': 'bar'})
+            settings_manager.set('foo', 'FOO', deprecated_msg='what to do instead')
+        assert 'what to do instead' in caplog.text
 
-    # Should return the setting as a string when the env var is set
-    ('foo', 'FOO', None, {'FOO': 'bar'}, {'foo': 'bar'}),
-    ('foo.bar', 'FOO', None, {'FOO': 'baz'}, {'foo.bar': 'baz'}),
+    @pytest.mark.parametrize('config,setting,varname,type_,environ,required,default,expected', (
+        # Should leave value at None when the env var in question isn't set
+        ({'foo': None}, 'foo', 'FOO', str, {}, False, None, None),
 
-    # Should coerce the result using the passed type
-    ('foo', 'FOO', asutf8, {'FOO': 'bar'}, {'foo': b'bar'}),
-    ('app_port', 'PORT', int, {'PORT': '123'}, {'app_port': 123}),
-))
-def test_env_setting(setting, varname, type, environ, expected):
-    func = settings.EnvSetting(setting, varname, type)
+        # Should return the setting as a string when the env var is set
+        ({}, 'foo', 'FOO', str, {'FOO': 'bar'}, False, None, 'bar'),
+        ({'foo.bar': 'foo.bar'}, 'foo.bar', 'FOO', str, {'FOO': 'baz'}, False, None, 'baz'),
 
-    result = func(environ)
+        # Should coerce the result using the passed type
+        ({}, 'foo', 'FOO', asutf8, {'FOO': 'bar'}, False, None, asutf8('bar')),
+        ({}, 'app_port', 'PORT', int, {'PORT': '123'}, False, None, 123),
 
-    assert result == expected
+        # Should overide the value when a default is provided
+        ({}, 'bar', 'BAR', str, {}, True, 'bar', 'bar'),
 
+        # Should not overide the value when a default is provided and the envvar is set
+        ({}, 'foo', 'FOO', str, {'FOO': 'bar'}, True, 'boo', 'bar'),
 
-def test_env_setting_returns_nicer_error_for_type_failure():
-    func = settings.EnvSetting('port', 'PORT', type=int)
+        # Should not error when required and a default is already set
+        ({'foo': 'foo'}, 'foo', 'FOO', str, {}, True, None, 'foo'),
 
-    with pytest.raises(settings.SettingError):
-        func({'PORT': 'notanint'})
+        # Should not overide default if default is already set
+        ({'boo': 'boo'}, 'boo', 'BOO', str, {}, True, 'foo', 'boo'),
 
+    ))
+    def test_env_setting(self, config, setting, varname, type_, environ, required, default, expected):
```

Comment:
Here we seem to be using the name `config` for the `settings` argument? I think it's confusing to change names around like that. I can see that the name `settings` would conflict with the imported `settings` module. Maybe the test should just do `from h.settings import SettingsManager`?

Reply:
`setting` is used for the `name` param here too